src/storage.cpp

```diff
@@ -103,11 +103,18 @@ namespace libtorrent {
 
 	void default_storage::need_partfile()
 	{
+		// even if std::call_once is used below, having this check
+		// avoid unnecessary use of locking machinery
 		if (m_part_file) return;
 
-		m_part_file.reset(new part_file(
-			m_save_path, m_part_file_name
-			, files().num_pieces(), files().piece_length()));
+		// the need of this synchronization is that writev runs in
+		// a multithreaded environment with the mutex unlocked
+		std::call_once(m_part_file_flag, [this]()
+		{
+			m_part_file.reset(new part_file(
```

Comment:
you would also need to reset the ``m_part_file_flag`` every time ``m_part_file`` is reset too, right?
it's also not obvious that the access to ``m_part_file`` in the if-statement above is race-free. There's no atomic operation when accessing it, but it may still be written to concurrently. I would imagine it would need to be a ``std::atomic<std::unique_ptr<>>`` or something.

Reply:
true, I didn't see that other `reset` in `delete_files`.

For what I read `std::call_once` is thread safe, could it be written concurrently?

Reply:
I think I understand your point better now, what are you saying is that even calling `.reset()` could be problematic, in the sense that counts as a write? am I right?